src/coreclr/jit/codegenarm64test.cpp

```diff
@@ -5912,6 +5912,42 @@ void CodeGen::genArm64EmitterUnitTestsSve()
     theEmitter->emitIns_R_R_R_I(INS_sve_udot, EA_SCALABLE, REG_V7, REG_V8, REG_V3, 3,
                                 INS_OPTS_SCALABLE_H); // UDOT <Zda>.S, <Zn>.H, <Zm>.H[<imm>]
 
+    // IF_SVE_EJ_3A
+    theEmitter->emitIns_R_R_R_I(INS_sve_cdot, EA_SCALABLE, REG_V0, REG_V1, REG_V2, 0,
```

Comment:
For these instructions that take a rotation value, shouldn't we be passing 0, 90, 180, 270 instead of 0, 1, 2, 3?

 I also implemented a few instructions that needed rotation values here: https://github.com/dotnet/runtime/pull/98141 which I am passing 0, 90, 180, 270.

Reply:
It looks a little weird passing 0-3 instead of 0-270, but I did this to match the bit-level representation of the rotation value, so that we don't need a helper method to encode the `rr` bits; then when displaying the instruction in JitDisasms, I multiply the immediate by 90 to display it correctly. I'm fine changing my approach to match yours. @kunalspathak @a74nh do you have any preference?

Reply:
The way I look at is the `emitIns` functions are APIs that should try to match what the instructions actually are. The bit-level representation/encoding is an implementation detail.

Reply:
I see. In that case, how about I wait for #98141 to be merged in, and then I'll update my encodings that use rotation values to use the helper methods you added?

Reply:
That seems fair to me

Reply:
I think theres a valid case for the `instr` to use only a small constant for the encoding, as that will save on memory usage.

When it's just two values I still think it makes sense to pass the full value into `emitIns_R_R_R_I()` and then encode as a single bit inside the function. 

For anything above 2 values, it can't fit into a single bit, so it needs to encode as an immediate. The min and max values for a small constant are -256 and 255. 270 doesn't fit into that range so we'd end up with more memory use.

So, it makes sense to me to pass 0, 90, 180, 270 to `emitIns_R_R_R_I()`, but to encode to 0-3 inside it.

(I'm also happy for us to use the non-optimal versions for now, and fix up later.)

@kunalspathak  ?

Reply:
That reminds me that some of the encodings have a two-bit rotation value and an immediate value; for those cases, I saved both values in the `instrDesc`'s immediate attribute (with the rotation value encoded as 0-3), and then later when we need the separate values, I used some bitwise operations to get them; saving the rotation value as two bits makes these bitwise operations easier, in my opinion.

> So, it makes sense to me to pass 0, 90, 180, 270 to emitIns_R_R_R_I(), but to encode to 0-3 inside it.

I agree, I'm fine with changing my approach to this.

Reply:
> So, it makes sense to me to pass 0, 90, 180, 270 to emitIns_R_R_R_I(), but to encode to 0-3 inside it.

I agree too

Reply:
I don't see a reason why we can't just have the values on `instrDesc` and then we encode them as 0, 1, 2, 3 at the actual time of encoding in `emitOutputInstr`

Reply:
> I don't see a reason why we can't just have the values on `instrDesc` and then we encode them as 0, 1, 2, 3 at the actual time of encoding in `emitOutputInstr`.


because of what @a74nh  pointed above:

>  The min and max values for a small constant are -256 and 255. 270 doesn't fit into that range so we'd end up with more memory use.

Reply:
To add to that, for encodings with an immediate and a rotation value, we squish these two values into the `instrDesc`'s constant field, so rotation values smaller than 270 can still result in us using the larger `instrDescCns` type.

Reply:
So, we want to allocate a smaller descriptor in which case values 0, 1, 2, 3 would need to be used. I guess that's fine, but I don't see a situation where memory would be an issue here, as these instructions are going to be rare.

Reply:
I would prefer we write for readability first if we know that such an optimization would not make any difference in 99.9% of scenarios.

However, I'm fine with encoding it as 0, 1, 2, 3 on `instrDesc` if that is what we all want. I'll have to adjust my work as well.

Reply:
> I would prefer we write for readability first

It is readable from the perspective of calling the `emitIns*` method as seen here: https://github.com/dotnet/runtime/pull/98187/files#diff-d4f9f119d0a321cea7e82023cb754d8abdb800d6185c8bb9464d389ebd50debcR6288

and we flip it just before saving it in instrdesc:

https://github.com/dotnet/runtime/pull/98187/files#diff-2b2c8b9011607926410624d6f81613fad7b74c6e0516d578675a8b792998fe4fR11110

I am not sure if `emitOutputInstr()` method is readable anyway :)

Reply:
It isn't necessarily about making `emitOutputInstr` readable, but about the display code being simple. We will have to decode the `imm` whose values are 0-3 to be translated to 0-270 on display.

Reply:
I guess what I'm trying to say is, if we encode the values as 0-3 on the `instrDesc`, we will have to have an `encode`/`decode` for it, whereas if we store 0-270, we only need one `encode` function.